File: typescript/src/domain/quantity.ts
/**
 * Quantity value object: a strictly positive integer.
 */

import { InvalidOrder } from "./errors";

/**
 * Shared maximum quantity (inclusive); 32-bit signed integer max.
 */
export const QUANTITY_MAX = 2_147_483_647;

/**
 * An amount of stock that must be strictly positive.
 *
 * Create instances through {@link Quantity.create}; the constructor is
 * private so no unvalidated quantity can exist.
 */
export class Quantity {
  /**
   * Creates a validated `Quantity`, raising `InvalidOrder` on violation.
   *
   * @param value - the raw integer amount; zero, negatives, fractions, and
   * values above {@link QUANTITY_MAX} are invalid.
   * @returns the validated value object.
   * @throws InvalidOrder when the value is not a strictly positive integer
   * in the shared range.
   */
  static create(value: number): Quantity {
    if (!Number.isSafeInteger(value) || value <= 0) {
      throw new InvalidOrder(`quantity must be a strictly positive integer, got ${String(value)}`);
    }
    if (value > QUANTITY_MAX) {
      throw new InvalidOrder(`quantity exceeds ${String(QUANTITY_MAX)}, got ${String(value)}`);
    }
    return new Quantity(value);
  }

  /**
   * The validated strictly positive integer amount.
   */
  readonly value: number;

  private constructor(value: number) {
    this.value = value;
  }
}

File: swift/Sources/Warehouse/Domain/Quantity.swift
/// Shared quantity maximum (`Int32.max`).
public let quantityMax = 2_147_483_647

/// An amount of stock that must be strictly positive.
public struct Quantity: Equatable, Hashable, Sendable {
  public let value: Int

  public init(_ value: Int) throws {
    if value <= 0 {
      throw InvalidOrder("quantity must be strictly positive, got \(value)")
    }
    if value > quantityMax {
      throw InvalidOrder("quantity exceeds \(quantityMax), got \(value)")
    }
    self.value = value
  }
}
